internal/transport/transport_test.go

```diff
@@ -2450,3 +2452,56 @@ func TestConnectionError_Unwrap(t *testing.T) {
 		t.Error("ConnectionError does not unwrap")
 	}
 }
+
+func (s) TestPeerSetInServerContext(t *testing.T) {
+	// create client and server transports.
+	server, client, cancel := setUp(t, 0, math.MaxUint32, normal)
+	defer cancel()
+	defer server.stop()
+	defer client.Close(fmt.Errorf("closed manually by test"))
+
+	// create a stream with client transport.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	stream, err := client.NewStream(ctx, &CallHdr{})
+	if err != nil {
+		t.Fatalf("failed to create a stream: %v", err)
+	}
+
+	// verify peer is set in client transport context.
+	if _, ok := peer.FromContext(client.ctx); !ok {
+		t.Fatalf("Peer expected in client transport's context, but actually not found.")
+	}
+
+	// verify peer is set in stream context.
+	if _, ok := peer.FromContext(stream.ctx); !ok {
+		t.Fatalf("Peer expected in stream context, but actually not found.")
+	}
+
+	// verify peer is set in server transport context.
+	count := 0
+	for {
```

Comment:
You could use the `waitWhileTrue` helper method to simplify the waiting logic here. See : https://github.com/grpc/grpc-go/blob/master/internal/transport/transport_test.go#L1616